### ISSUE TITLE
vec_init_then_push incorrectly fires when the `push()` is in a `loop` body

### ISSUE THREAD

**Issue by duci9y** [rust-lang/rust-clippy#7019]
Lint name: vec_init_then_push

I tried this code:

```rust
let mut v = Vec::new();
loop {
    match *s {
        ...
        SExp::Cons(ref a, ref b) => {
            v.push(a);
        }
        ...
    }
}
```

I expected to see this happen: The lint to not fire, since this is a valid use case for init then push.

Instead, this happened: The lint fired.

```rust
378 | /         let mut v = Vec::new();
379 | |         loop {
380 | |             match *s {
381 | |                 ...
382 | |                 SExp::Cons(ref a, ref b) => {
383 | |                     v.push(a);
    | |______________________________^ help: consider using the `vec![]` macro: `let mut v = vec![..];`
```

### Meta

- `cargo clippy -V`: clippy 0.1.51 (2fd73fab 2021-03-23)
- `rustc -Vv`:
```
rustc 1.51.0 (2fd73fabe 2021-03-23)
binary: rustc
commit-hash: 2fd73fabe469357a12c2c974c140f67e7cdd76d0
commit-date: 2021-03-23
host: x86_64-apple-darwin
release: 1.51.0
LLVM version: 11.0.1
```

**Comment by camsteffen** [comment]
This should be fixed by #6697. Can you provide a minimal reproducible example?

**Comment by duci9y** [comment]
I spent about 15 minutes trying to get a repro going but couldn't trigger the false positive. I'll try again later or see what I can gather about the context.

**Comment by TyPR124** [comment]
I came across this, and managed to simplify syntax down to this. I can reproduce locally `clippy 0.1.51 (2fd73fa 2021-03-23)`, but not on play.rust-lang.org `2021-04-27 727d101` (as of writing, true for both this example and my real code).

```rust
#[allow(clippy::while_let_loop)]

fn foo(a: &[()]) {
    let mut v = vec![];
    loop {
        match a.iter().next() {
            Some(_) => {
                v.push(());
            }
            _ => break,
        }
    }
}
```

**Comment by camsteffen** [comment]
> I can reproduce locally `clippy 0.1.51 (2fd73fa 2021-03-23)`

That version predates #6697. Your example no longer lints.

**Comment by TyPR124** [comment]
That's why I mentioned play.rust-lang.org as well, as I assume that version included that fix. Is there something that prevents that example from being a MCVE?

**Comment by camsteffen** [comment]
We need an example that is a false positive in the latest version, or else there is no issue here.

**Comment by TyPR124** [comment]
But the issue was opened with for clippy 0.1.51 (2fd73fab 2021-03-23), the same version I was using.

> or else there is no issue here.

Yes, my example leads me to believe this issue has been resolved.
